packages/web/src/components/apps/use-preview.tsx

```diff
@@ -43,14 +46,25 @@ export function PreviewProvider({ channel, children }: ProviderPropsType) {
     return () => channel.off('preview:status', onStatusUpdate);
   }, [channel, setUrl, setStatus]);
 
-  function start() {
+  async function start() {
+    if (nodeModulesExists === false) {
+      await npmInstall();
+    }
     channel.push('preview:start', {});
   }
 
   function stop() {
     channel.push('preview:stop', {});
   }
 
+  // If the node_modules directory gets deleted, then stop the preview server
+  useEffect(() => {
+    if (nodeModulesExists !== false) {
+      return;
+    }
+    stop();
+  }, [nodeModulesExists]);
+
   // When the page initially loads, start the vite server
```

Comment:
If the `node_modules` directory gets deleted, the preview server will automatically stop.